### DISCUSSION TITLE
Recipe Variants

### DISCUSSION THREAD

**Discussion by jakeanq** [cooklang/spec#75] [Ideas]
There are often recipes which are variations on a theme, with broadly similar preparation methods and it would be helpful to support these.  For simple cases (eg substituting a different type of fruit) it might be enough to use optional ingredients (#50) and alternative ingredients (#68), however for more complex differences this wouldn't work so well.  There is also the question of specifying which options you want to use - sometimes it could be controlled by the shopping list config (eg if you always want the vegan version, leave animal products off of the shopping list config), but if the only difference is blackberries vs blueberries, that's not so good.

As an initial syntax proposal, how about a variation tag (or tags) in square brackets.

```
[apricot,cranberry] Soak @dried apricot[apricot]|cranberry[cranberry]{1 cup} in water for 1 hour. -- Ingredient options may be specified for each option separately (using suggested syntax from #50 )

Add @flour{3 cups} and @dry yeast{5g} to a bowl. -- All variations

[olive] Drain @green olives{1/2 cup}, chop and add to bowl. -- Entire step is only for the olive variation; ingredients specified in this step become tied to the tag.

[apricot,cranberry] Drain fruit and add to bowl. -- Step is valid for either of the variations

Season with @pepper[olive]|cinnamon[apricot]|powdered ginger[cranberry]{1tbsp}.

-- Insert rest of bread recipe here...
```

When preparing a shopping list or viewing the recipe etc in the CLI, the variation could be selected by using a command-line flag or URL fragment, for example:
```bash
cook shopping-list "bread.cook#olive" "bread.cook#apricot" # Probably the preferred option
# or
cook shopping-list bread.cook --variant olive # Argument would presumably apply to the previously-specified file.
```
In GUIs, presumably a drop-down menu or similar would be available.  Maybe there should be a standardised way to specify "pretty" names too?

**Comment by ziofat** [comment]
Really like this.

**Comment by chrbala** [comment]
From what I have considered, there are two broad categories of variants:
1. Variants which change the intended outcome of the recipe. This could be different ingredients or preparation methods. 
2. Variants which adjust the preparation based on available tools, ingredients, or desired effort from the cook. Functionally is more along the lines of substitutions, but it's still very similar to the first category. 

On an unrelated note, my own preference would be to have variants require an explanation to why the variant might be chosen. It's frustrating to see a recipe which says a particular step or ingredient is optional: is it worth it for me to do or not? Some may be fairly obvious: "add raisins to your muffins if you want a sweeter muffin". Others may be less clear.

**Comment by shicks** [comment]
The use cases I have mind would be relatively generic recipes that serve as the base for a number of different possibilities.  For instance, we cook a simple "rice and beans" dish that can go with pinto beans, kidney beans, or black beans - but we use different seasonings for each.  Or a cookie recipe base that can make chocolate chip, oatmeal raisin, peanut butter, etc, by adding different extra ingredients at various stages.

**Comment by bubonicfred** [comment]
I was just thinking about this and came to the same conclusion as the first post

**Comment by keithamus** [comment]
Some recipes also can substitute multiple ingredients for one, for example curries might be able to substitute spices and sauces for a pre-made curry base, or cocktails might substitute citrus+syrup for a sour mix which some bars might have on hand. 

It can also be useful to have more "generic" recipes, which provide the base set of instructions for a wide variety of dishes. As an example here's a generic "collins" base recipe which (although very basic) demonstrates how recipes could expand into dozens of concrete varieties:

```
Fill a #collins glass{} with @ice{100%g}.

Add @simple syrup{15%ml}, @lemon juice{30%ml}, @gin{45%ml} and stir with a #bar spoon{}.

Top with @carbonated water{60%ml}, and gently stir.

Garnish with a @lemon{5%g}(1 slice) and @maraschino cherries{5%g}.

> For a "Tom Collins" one must use "Old Tom Gin".

> A "John Collins" traditionally calls for Gin, however more commonly a John Collins uses Rye Whiskey instead of Gin.

> "Juan Collins" or "Jose Collins" calls for Tequila instead of Gin.

> "Comrade Collins" calls for Vodka, instead of Gin.

> "Michael Collins" calls for Irish whiskey, instead of Gin.

> "Sandy Collins" calls for Scotch whiskey, instead of Gin.

> "Russell Collins" calls for Jagermeister, instead of Gin.

> "Ron Collins" calls for Rum, instead of Gin.

> If you have sour-mix then 30 ml of this can replace the lemon juice and simple syrup.
```

**Comment by tmlmt** [comment]
I've implemented recipe variants in my own TS parser. See [brief guide](https://cooklang-parser.tmlmt.com/v3/guide-extensions.html#recipe-variants) and [playground](https://cooklang-parser.tmlmt.com/playground/). Feedback is welcome :)
